src/array_api_extra/testing.py

```diff
@@ -0,0 +1,248 @@
+from types import ModuleType
+from typing import TYPE_CHECKING, Any, TypeVar, cast
+
+import pytest
+
+from array_api_extra._lib._utils._compat import is_dask_namespace, is_jax_namespace
+
+__all__ = ["lazy_xp_function", "patch_lazy_xp_functions"]
+
+if TYPE_CHECKING:
+    # TODO move ParamSpec outside TYPE_CHECKING
+    # depends on scikit-learn abandoning Python 3.9
+    # https://github.com/scikit-learn/scikit-learn/pull/27910#issuecomment-2568023972
+    from typing import ParamSpec
+
+    from dask.typing import Graph, Key, SchedulerGetCallable
+    from typing_extensions import override
+
+    P = ParamSpec("P")
+else:
+    SchedulerGetCallable = object
+
+    # Sphinx hacks
+    class P:  # pylint: disable=missing-class-docstring
+        args: tuple
+        kwargs: dict
+
+    def override(func: Callable[P, T]) -> Callable[P, T]:
+        return func
+
+
+T = TypeVar("T")
+
+
+def lazy_xp_function(  # type: ignore[no-any-explicit]
+    func: Callable[..., Any],
+    *,
+    allow_dask_compute: int = 0,
+    jax_jit: bool = True,
+    static_argnums: int | Sequence[int] | None = None,
+    static_argnames: str | Iterable[str] | None = None,
+) -> None:  # numpydoc ignore=GL07
+    """
+    Tag a function to be tested on lazy backends.
+
+    Tag a function, which must be imported in the test module globals, so that when any
+    tests defined in the same module are executed with `xp=jax.numpy` the function is
+    replaced with a jitted version of itself, and when it is executed with
+    `xp=dask.array` the function will raise if it attempts to materialize the graph.
+    This will be later expanded to provide test coverage for other lazy backends.
+
+    In order for the tag to be effective, the test or a fixture must call
+    :func:`patch_lazy_xp_functions`.
+
+    Parameters
+    ----------
+    func : callable
+        Function to be tested.
+    allow_dask_compute : int, optional
+        Number of times `func` is allowed to internally materialize the Dask graph. This
+        is typically triggered by `bool()`, `float()`, or `np.asarray()`.
+
+        Set to 1 if you are aware that `func` converts the input parameters to numpy and
+        want to let it do so at least for the time being, knowing that it is going to be
+        extremely detrimental for performance.
+
+        If a test needs values higher than 1 to pass, it is a canary that the conversion
+        to numpy/bool/float is happening multiple times, which translates to multiple
+        computations of the whole graph. Short of making the function fully lazy, you
+        should at least add explicit calls to `np.asarray()` early in the function.
+        *Note:* the counter of `allow_dask_compute` resets after each call to `func`, so
+        a test function that invokes `func` multiple times should still work with this
+        parameter set to 1.
+
+        Default: 0, meaning that `func` must be fully lazy and never materialize the
+        graph.
```

Comment:
In a previous version this was a all-or-nothing test. https://github.com/scipy/scipy/pull/22240#discussion_r1925408361 prompted me to change it.